luigi/contrib/dropbox.py

```diff
@@ -0,0 +1,275 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2019 Jose-Ignacio Riaño Chico
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not
+# use this file except in compliance with the License. You may obtain a copy of
+# the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+#
+
+from __future__ import absolute_import
+
+import datetime
+import logging
+import ntpath
+import os
+import random
+import tempfile
+from contextlib import contextmanager
+
+import luigi.format
+from luigi.target import FileSystem, FileSystemTarget, AtomicLocalFile
+
+logger = logging.getLogger('luigi-interface')
+
+try:
+    import dropbox.dropbox
+    import dropbox.exceptions
+    import dropbox.files
+except ImportError:
+    logger.warning("Loading Dropbox module without the python package dropbox (https://pypi.org/project/dropbox/). "
+                   "Will crash at runtime if Dropbox functionality is used.")
+
+
+class DropboxClient(FileSystem):
+    """
+    Dropbox client for authentication, designed to be used by the :py:class:`DropboxTarget` class.
+    """
+
+    def __init__(self, token):
+        """
+        :param str token: Dropbox Oauth2 Token. See :class:`DropboxTarget` for more information about generating a token
+        """
+        if not token:
+            raise ValueError("The token parameter must contain a valid Dropbox Oauth2 Token")
+
+        try:
+            conn = dropbox.dropbox.Dropbox(oauth2_access_token=token, user_agent="Luigi")
+        except Exception as e:
+            raise Exception("Cannot connect to Dropbox. Check your Internet connection and the token. \n" + repr(e))
+
+        self.token = token
+        self.conn = conn
+
+    def exists(self, path):
+        if path == '/':
+            return True
+        try:
+            md = self.conn.files_get_metadata(path)
+            return bool(md)
+        except dropbox.exceptions.ApiError as e:
+            if isinstance(e.error.get_path(), dropbox.files.LookupError):
+                return False
+            else:
+                raise e
+
+    def remove(self, path, recursive=True, skip_trash=True):
+        if not self.exists(path):
+            return False
+        self.conn.files_delete_v2(path)
+        return True
+
+    def mkdir(self, path, parents=True, raise_if_exists=False):
+        if self.exists(path):
+            if not self.isdir(path):
+                raise luigi.target.NotADirectory()
+            elif raise_if_exists:
+                raise luigi.target.FileAlreadyExists()
+            else:
+                return
+
+        self.conn.files_create_folder_v2(path)
```

Comment:
Just checking... if i provided `/this/is/my/path`, the dir would be created - same as if i provided `/this/is/my/path/` (assuming the dir `/this/is/my/path/` nor the file `/this/is/my/path` didn't already exist)

Reply:
In Dropbox API, trying to create a folder named '/this/is/my/path/' raises an "dropbox.expceptions.ApiError (malformed_path)", because a dropbox path cannot end with /.

We can hide this fact from the user (removing the trailing slash whenever he uses it) or we can only let him use the paths in the format that the API expects.

I have opted for the later option, but I am open to discuss it.

Reply:
Interesting. Our s3 module does the opposite, it ensures that the mkdir path ends with a trailing slash. Perhaps we should allow the input of a trailing slash path to this method, but strip it if it exists?

Reply:
I have been thinking about allowing a trailing slash path to this method, but I think it might cause some unexpected issues:

`DropboxClient.mkdir('/this/is/my/path/')` would work as the developer might expect (i.e. it creates the `/this/is/my/path`), but unfortunately `DropboxClient.exists('/this/is/my/path/')` would return False, which would surprise the developer.

If we decide to also strip the trailing slash in the exists() method, then `DropboxClient.exists('/files/existing_file.txt/')` would return True if the existing_file.txt exists (even if it is not a directory), which would surprise the developer.


I think that the current approach would be one of minimum surprise, and therefore a good one:
- Dropbox API does not allow the creation of paths ending with '/', so DropboxClient.mkdir('/') raises a ApiError due to `malformed_path` 
- Since no path ending with '/' can exist, `DropboxClient.exists(/existing/directory/')` returns false, and additionally logs a warning [_[link to code]_](https://github.com/joseignaciorc/luigi/blob/54586ed1d06a3576f53dd059e02ccbe736d00731/luigi/contrib/dropbox.py#L66 ). (Maybe it should be reworded? or changed into an exception to make it more obvious?).
- In DropboxTarget documentation, it is clearly stated that `The path must start with '/' and must not end with '/' (even if it is a directory).` [_[link to code]_](https://github.com/joseignaciorc/luigi/blob/master/luigi/contrib/dropbox.py#L220).

Reply:
I'm still struggling here. I think my difficulty is how unnatural it feels to not be permitted to mkdir with a trailing slash.

@Tarrasch or @honnix Could y'all chime in on this one? How flexible should we be with developers and trailing slashes for `mkdir` and `exists`, given Dropbox's lack of support for trailing slashes?

Reply:
@Tarrasch, @honnix: ping :-)

Reply:
I have been thinking about this. And, in order to give DroboxClient a sane behaviour with trailing-slash paths, I have created two decorators `accept_trailing_slash_in_existing_dirpaths` and `accept_trailing_slash`.

Now, `DropboxClient.mkdir('/this/is/my/path/')` works exactly as `DropboxClient.mkdir('/this/is/my/path')` and `DropboxClient.exists('/files/foo.bar/')`returns `True` only if '_/files/foo.bar/_' exists and is a directory. 

The rest of functions (e.g. `DropboxClient.copy`, `DropboxClient.move`) also have an analogous behavior, accepting trailing slashes for already-existing directories.

Additionaly if the developer tries to use a path ending with a slash, a warning (by `logger.warning`) is issued, because checking if a directory exists (and therefore a trailing-slash should be allowed) requires an additional Dropbox-API call.

I have added several unit tests to verify the behavior of this paths. Maybe the most interesting ones are the [dir_lifecycle](https://github.com/joseignaciorc/luigi/blob/1276a26eea80bbfba4460b0948b90b2d9e0cdf2f/test/contrib/dropbox_test.py#L158) tests.